### PR TITLE
Improve with providing Nuclear Norm Constraint

### PR DESCRIPTION
**Pull request by pkan2**
Hello!

I have updated the conditional_gradient optimizer with providing the option of nuclear norm constraint option, as described in issue #1105,

with using following formula:

```
variable -= (1-learning_rate) * (variable
    + lambda_ * top_singular_vector(gradient))
```

Thank you!

### CODE REVIEW

**Comment by pkan2** [comment]
Hello @facaiy @WindQAQ @Squadrick! 

Can you provide some feedback for the update of CG optimizer? 

Thank you!

File: tensorflow_addons/optimizers/conditional_gradient_test.py [hunk 1]
```diff
@@ -123,18 +150,158 @@ def lambda_():
                 self.evaluate(var1),
             )
 
-    def testBasic(self):
+    def doTestBasicNuclear(self, use_resource=False, use_callable_params=False):
+        # for i, dtype in enumerate([tf.half, tf.float32, tf.float64]):
```
Comment (Squadrick) [inline tensorflow_addons/optimizers/conditional_gradient_test.py]: Delete this line. Or check for windows, and selectively exclude `tf.half`.

**Comment by gabrieldemarmiesse** [comment]
@googlebot I consent

**Comment by gabrieldemarmiesse** [comment]
I've merged master into your branch to update it and fixed any formatting/conflicts it might have. If you need to do some more modifications, please do `git pull` beforehand.

**Comment by pkan2** [comment]
@gabrieldemarmiesse Got it! Thank you a lot!

**Comment by gabrieldemarmiesse** [comment]
@Squadrick when you have the time, could you look into this pull request? Thank you.

File: tensorflow_addons/optimizers/conditional_gradient_test.py [hunk 2]
```diff
@@ -26,7 +27,63 @@ def _update_conditional_gradient_numpy(self, var, norm, g, lr, lambda_):
         var = var * lr - (1 - lr) * lambda_ * g / norm
         return var
 
-    def doTestBasic(self, use_resource=False, use_callable_params=False):
+    def top_singular_vector(self, m):
```
Comment (Squadrick) [inline tensorflow_addons/optimizers/conditional_gradient_test.py]: You can instead use the method previously defined: `cg_lib.ConditionalGradient._top_singular_vector`. So if any changes are made to the actual impl will reflect here.

File: tensorflow_addons/optimizers/conditional_gradient.py [hunk 3]
```diff
@@ -131,23 +193,31 @@ def frobenius_norm(m):
...
+        else:
+            top_singular_vector = tf.convert_to_tensor(
+                self._top_singular_vector(grad),
+                name="top_singular_vector",
+                dtype=var.dtype.base_dtype,
+            )
+            var_update_value = (
+                tf.math.multiply(var_slice, lr)
+                - (1 - lr) * lambda_ * top_singular_vector
+            )
```
Comment (Squadrick) [inline tensorflow_addons/optimizers/conditional_gradient.py]: The `var_update_value` in both branches have common terms:
```
tf.math.multiply(var_slice, lr) - lambda_ * (1 - lr)
```

You can rewrite this as:

```
factor = tf.math.multiply(var_slice, lr) - lambda_ * (1 - lr)

if self.ord == "fro":
  norm = tf.convert_to_tensor(...)
  var_update_value = factor * grad / (norm + eps)
else:
  top_singlar_vector = tf.convert_to_tensor(...)
  var_update_value = factor * top_singular_vector
```

Makes the code easier to read.

Comment (Squadrick) [reply tensorflow_addons/optimizers/conditional_gradient.py]: This happens in `_resource_apply_dense` as well.

File: tensorflow_addons/optimizers/conditional_gradient_test.py (hunk 2 above)
Comment (pkan2) [reply tensorflow_addons/optimizers/conditional_gradient_test.py]: Got it! Thank you!

**Comment by Squadrick** [comment]
@pkan2 There's a merge conflict, can you fix that? Once it's done, this should get merged.

**Comment by pkan2** [comment]
Hello @Squadrick.Thank you for your feedback! I have fixed the merging conflict. 

Thank you!

**Comment by gabrieldemarmiesse** [comment]
I'll merge this as it was approved by @Squadrick

**Pull request merged**